### PR TITLE
Outside via macvlan

### PR DESCRIPTION
**Pull request by iHalt10**
**What this PR does / why we need it**:
I have taken over this PR. If you have any problems, please tell me.
> https://github.com/kubevirt/kubevirt/pull/2192

When using multus to start a VM with macvlan, get an ip, but there is a problem that cannot communicate with the outside.
This happened because of a configuration issue with the linux bridge Forwarding Database.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://github.com/kubevirt/kubevirt/issues/2925

**Special notes for your reviewer**:

**Release note**:

```release-note
Outside via macvlan
```

### CODE REVIEW

**Comment by iHalt10** [comment]
I haven't done DCO signoff on the original PR code that I took over.
Is it better not to merge the original PR code ... !?

File: pkg/virt-launcher/virtwrap/network/common.go [hunk 1]
```diff
@@ -403,3 +404,9 @@ func setInterfaceCacheFile(path string) {
 func setVifCacheFile(path string) {
 	vifCacheFile = path
 }
+
+func (h *NetworkUtilsHandler) DelBridgeFdb(iface, macAddress string) error {
```
[the remark this answers is not in the document]
Comment (iHalt10) [reply pkg/virt-launcher/virtwrap/network/common.go:408]: I thought about it easily, but I will change it to use netlink.

File: tests/vmi_multus_test.go [hunk 2]
```diff
@@ -311,6 +312,150 @@ var _ = Describe("Multus", func() {
...
+				)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(out).To(Equal("0\n"))
+			},
+				table.Entry("when not using tuning plugin", "ptp-conf"),
+				table.Entry("when using tuning plugin", "ptp-conf-tuning"),
+			)
+		})
+
+		Context("VirtualMachineInstance with sriov plugin interface", func() {
```
[the remark this answers is not in the document]
Comment (iHalt10) [reply tests/vmi_multus_test.go:339]: I’m sorry. it seems that when the original PR branch was merged, the old code was merged as well.
I'll delete it.

**Comment by DanielQujun** [comment]
@iHalt10 Hi, I try to verify it with some Linux commands, but without luck, could you please have a look? 

discuss group: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/kubevirt-dev/8_VsGdWlVaE

**Comment by iHalt10** [comment]
@DanielQujun Hi , I also confirmed that I could communicate with the commands described in "discuss group".

In the "iHalt10:outside_via_macvlan" branch, I implemented it by executing it from "golang netlink api" instead of from the command. It can communicate.

However, since there is "macvtap" in addition to "macvlan", there is not much difference from it.
The implementation of "macvtap" can use "dhcpd" inside the container of virt-launcher.

**Comment by iHalt10** [comment]
So I'm waiting for the "macvtap" implementation to finish.
Then, it is better to clarify the difference between "macvlan" and "macvtap".

If you want to make the communication with the outside work immediately, you may execute it in this "iHalt10:outside_via_macvlan" branch.

**Comment by DanielQujun** [comment]
> @DanielQujun Hi , I also confirmed that I could communicate with the commands described in "discuss group".

Hi, Appreciate for you confirm, I have been spending hours on it:sob:, so if the commands work for you, there might be some defect in my environment.

for some reason, my CNI can not add macvlan interface directly, so I use Linux commands to do a POC.

**Comment by DanielQujun** [comment]
Hi, I'm not familiar with L2 networks, It's hard for me to figure out why we delete the pod interface connected to the bridge from the forward database of the linux bridge, the macvlan Interface would forward the package to bridge, then the VMI could be able to communicate outside? 

would you pls elaborate more or post some related links?

**Comment by DanielQujun** [comment]
according to this posts:[Conflicts and Limitations of Bridge and Macvlan Devices
](https://cnly.github.io/2018/11/09/conflicts-and-limitations-of-bridge-and-macvlan-devices.html)

It looks macvlan interface can not be subinterface of a bridge, did I miss something?

**Comment by iHalt10** [comment]
Ah, I did not execute it with the all same command as below.
```
  1.add Macvlan on host:
       `ip link add dcs21 link eth0 type macvlan mode bridge`
  2.enable promisc mode for host eth0:
       `ifconfig eth0 promisc`
  3.set Macvlan Interface into VMI's POD:
     `ip link set dcs2 netns dcs2 up`
  4.add a new bridge in the VMI's POD
      `brctl addbr macvlanbr`
  5. set that bridge up:
      `ip link set macvlanbr up`
  6.set macvlan as a subinterface of bridge:
      `brctl addif macvlanbr dcs2`
  7.delete macvlan's MAC from fdb
      `bridge fdb delete 32:58:08:96:34:04 dev dcs2 master`
  8. attach that bridge network into VMI
      `virsh attach-interface 1 --type bridge --source macvlanbr`
  9.set link up and add IP address for the new interface in VMI
     `ip link set eth1 up && ip add add 192.168.0.20/16 dev eth1 `
```
In fact, what I did was
```
  7.delete macvlan's MAC from fdb
      `bridge fdb delete 32:58:08:96:34:04 dev dcs2 master
and
Change VM network interface mac address
```

**Comment by iHalt10** [comment]
Yes. I can explain.
However, it is difficult to explain, so it would be helpful if you could wait a few hours.
(The behavior of the L2 that implemented is extremely bizarre. )

**Comment by EdDev** [comment]
> However, since there is "macvtap" in addition to "macvlan", there is not much difference from it.

For clarification, the two options (macvlan & macvtap) are not the same and results in different setups:
- macvlan:  nodeNic0 - macvlan0 - podBridge0 - tap0 - vnic(vm)
- macvtap: nodeNic0 - macvtap0 - vnic(vm)

> The implementation of "macvlan" can use "dhcpd" inside the container of virt-launcher.

Yes, the macvlan will allow the masquerade binging with a DHCP server on the pod while the macvtap will not and will be available **only** on secondary networks/interfaces.

**Comment by iHalt10** [comment]
@DanielQujun 

I will explain the behavior of the implemented "macvlan".
This was due to a configuration issue with the "Linux Bridge Forwarding Database" and "vm mac addr".

The figure below shows the configuration that cannot communicate (currently master commit).
```
host-node                                  |  Physical Nic  |
                                               |       |
                                               |    | Logic ifce | mac[11:11:11:11:11]
                                               |
----------------------------------------------------------(netns)
virt-launcher pod:compute container:       | eth0(macvlan) | mac[96:1f:c6:bf:89:8c]
                                               |
        +-------| k6t-eth0(bridge) |-----------+
        |         bridge fdb show:
        |    (There is fdb for "96:1f:c6:bf:89:8c" dev eth0)
        |
        |
     | vnet0 | mac[fe:1f:c6:bf:89:8c]
        |
-------------------------------------------------------
VM:  | eth0  | (1): mac[22:22:22:22:22] => (2): mac[96:1f:c6:bf:89:8c]

```
(1):
VM cannot communicate because 'mac aadr of VM:eth0' and 'mac addr of container:eth0(macvlan)' are not the same. (Outbound works, but not Inbound)

So, considering 'VM:eth0' and 'container:eth0(macvlan)' as the same "mac addr", the following problems occur.
(2):
Outbound network packet from VM is forwarded to 'container:eth0(macvlan)' by bridge fdb. So, it is sent outside.
However, Inbound network packet from "Physical Nic" to "VM" does not reach the VM.
Inbound network packet reaches 'k6t-eth0' after passing "eth0(macvlan)".
And because "bridge fdb" knows to forward to "container:eth0(macvlan)", it is not forwarded to "vnet0-vm:eth0".


The figure below shows the configuration that enables communication.
```
host-node                                  |  Physical Nic  |
                                               |       |
                                               |    | Logic ifce | mac[11:11:11:11:11]
                                               |
----------------------------------------------------------(netns)
virt-launcher pod:compute container:       | eth0(macvlan) | mac[96:1f:c6:bf:89:8c]
                                               |
        +-------| k6t-eth0(bridge) |-----------+
        |         bridge fdb show:
        |    (There is no fdb for "96:1f:c6:bf:89:8c")
        |
        |
     | vnet0 | mac[fe:1f:c6:bf:89:8c]
        |
-------------------------------------------------------
VM:  | eth0  | mac[96:1f:c6:bf:89:8c]

```
Outbound network packet from the VM reaches k6t-eth0(bridge).
"bridge fdb" does not have a record of "96:1f:c6:bf:89:8c", so it forwards the packet to all ports connected to bridge.
So, it is forwarded to "container:eth0 (macvlan)" and sent outside.

This behavior is the same as the "repeater hub".

And in the same way, Inbound network packet also reaches k6t-eth0(bridge), forwards the packet to all ports connected to bridge, and reaches "vnet0->VM:eth0".

**Comment by DanielQujun** [comment]
> @DanielQujun
> 
> I will explain the behavior of the implemented "macvlan".
> This was due to a configuration issue with the "Linux Bridge Forwarding Database" and "vm mac addr".

Awesome! The magic conceals in that VMI's mac needs to be the same as "container:eth0(macvlan)", I did not know that, I used to thought the mac address should not be same. my network is working now after tried as you said. although there is a minor flaw that It seems got duplicates package on the VMI's eth0.
```
PING 192.168.0.18 (192.168.0.18) 56(84) bytes of data.
64 bytes from 192.168.0.18: icmp_seq=1 ttl=64 time=0.680 ms
64 bytes from 192.168.0.18: icmp_seq=1 ttl=63 time=0.681 ms (DUP!)
64 bytes from 192.168.0.18: icmp_seq=2 ttl=64 time=0.346 ms
64 bytes from 192.168.0.18: icmp_seq=2 ttl=63 time=0.346 ms (DUP!)
64 bytes from 192.168.0.18: icmp_seq=3 ttl=64 time=0.438 ms
64 bytes from 192.168.0.18: icmp_seq=3 ttl=63 time=0.438 ms (DUP!)
```
Anyway, I have got big progress :laughing:

@iHalt10 @EdDev Thank you for your professional explanations!

**Comment by iHalt10** [comment]
For the time being, I'll fix this branch in the current master branch.
Please wait.
After that, please review.

**Comment by cumirror** [comment]
> "bridge fdb" does not have a record of "96:1f:c6:bf:89:8c", so it forwards the packet to all ports connected to bridge.
> So, it is forwarded to "container:eth0 (macvlan)" and sent outside.

@iHalt10 after the fdb was deleted, i found bridge would learn mac to update fdb, like this:
```
[~]# sudo nsenter -t 21318 -n bridge fdb show k6t-net1 | grep f2:9a:17:a4:f0:99
f2:9a:17:a4:f0:99 dev vnet1 master k6t-net1
```
f2:9a:17:a4:f0:99 belong to vnet1 (macvlan dev).
this rule does not affect traffic forwarding, so is it necessary to close the bridge mac learning?

**Comment by iHalt10** [comment]
@cumirror 
Circling back to this, sorry for the delay.
Thank you for your comment.

You're right.
What I explained is flawed.
 > https://github.com/kubevirt/kubevirt/pull/3489#issuecomment-670198798

So I'll fix it.
"Bridge fdb is dynamically configured" instead of "hub-like behavior".

**Comment by gabrielecerami** [comment]
@iHalt10: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by DanielQujun** [comment]
I'm encountering a problem when doing tests on windows, It seems that windows won't send request when the DHCP server which running by virt-launcher have sent reply. 

```
root@test-hpa-0-0-1-2:~# tcpdump -i tap0 -nn -vv port 67
tcpdump: listening on tap0, link-type EN10MB (Ethernet), capture size 262144 bytes
14:15:53.632757 IP (tos 0x0, ttl 128, id 36519, offset 0, flags [none], proto UDP (17), length 328)
    0.0.0.0.68 > 255.255.255.255.67: [udp sum ok] BOOTP/DHCP, Request from 02:42:ae:af:00:05, length 300, xid 0x70303db0, secs 2560, Flags [none] (0x0000)
          Client-Ethernet-Address 02:42:ae:af:00:05
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Discover
            Client-ID Option 61, length 7: ether 02:42:ae:af:00:05
            Hostname Option 12, length 15: "WIN-9PF63QC6R1U"
            Vendor-Class Option 60, length 8: "MSFT 5.0"
            Parameter-Request Option 55, length 14: 
              Subnet-Mask, Default-Gateway, Domain-Name-Server, Domain-Name
              Router-Discovery, Static-Route, Vendor-Option, Netbios-Name-Server
              Netbios-Node, Netbios-Scope, Option 119, Classless-Static-Route
              Classless-Static-Route-Microsoft, Option 252
14:15:53.633033 IP (tos 0x0, ttl 64, id 48869, offset 0, flags [DF], proto UDP (17), length 416)
    169.254.75.10.67 > 255.255.255.255.68: [udp sum ok] BOOTP/DHCP, Reply, length 388, xid 0x70303db0, Flags [none] (0x0000)
          Your-IP 174.175.0.5
          Client-Ethernet-Address 02:42:ae:af:00:05
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Offer
            Server-ID Option 54, length 4: 169.254.75.10
            Lease-Time Option 51, length 4: 86313600
            MTU Option 26, length 2: 1500
            Classless-Static-Route Option 121, length 5: (default:174.175.0.1)
            T119 Option 119, length 65: 11.54.45.45.56.56.48.51.55.53.54.54.3.115.118.99.7.99.108.117.115.116.101.114.5.108.111.99.97.108.0.3.115.118.99.7.99.108.117.115.116.101.114.5.108.111.99.97.108.0.7.99.108.117.115.116.101.114.5.108.111.99.97.108.0
            Hostname Option 12, length 3: "aaa"
            Domain-Name Option 15, length 29: "6--88037566.svc.cluster.local"
            Subnet-Mask Option 1, length 4: 255.255.255.0
            Default-Gateway Option 3, length 4: 174.175.0.1
            Domain-Name-Server Option 6, length 4: 192.168.0.230
```

However, the Windows VM would send request back to the DHCP server which running in the LAN. So I think the problem might be caused by the same mac between VM and Bridge.

**Comment by DanielQujun** [comment]
Still failed to assign IP when Windows VM is using calico IPIP network... 

Wireshark in Windows could capture "DHCP Discover" and "DHCP Offer" packets, But the Event Viewer says "Offer Receive Timeout has happened".
wireshark-dhcp

I'm afraid that the might some mac filter rules drop the DHCP offer, cause everything is OK in Linux VM. But I have run out of ideas to debug the deep truth.:sob:

@iHalt10, Do you have a chance to test Windows?

**Comment by DanielQujun** [comment]
not sure whether is The right direction, I add a veth pair, make the DHCP server listen on one of it, and the other connects with bridge k6t-eth0, the windows could receive DHCP offer now.

```
root@test-hpa-0-0-1-2:~# docker exec -it 8357de00b9b2 bash
[root@windows-test-dhcp /]# ip add
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
3: k6t-eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:c0:a8:01:16 brd ff:ff:ff:ff:ff:ff
    inet 169.254.75.10/32 scope global k6t-eth0
       valid_lft forever preferred_lft forever
4: dhcp-peer@dhcp: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master k6t-eth0 state UP group default 
    link/ether 7e:dc:ed:5a:f2:5c brd ff:ff:ff:ff:ff:ff
5: dhcp@dhcp-peer: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 16:1d:1d:b0:66:a5 brd ff:ff:ff:ff:ff:ff
    inet 169.254.75.9/32 scope global dhcp
       valid_lft forever preferred_lft forever
6: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master k6t-eth0 state UP group default qlen 1000
    link/ether 42:7b:2a:d2:ba:f9 brd ff:ff:ff:ff:ff:ff
232: eth0@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master k6t-eth0 state UP group default 
    link/ether 02:42:c0:a8:01:16 brd ff:ff:ff:ff:ff:ff link-netnsid 0
[root@windows-test-dhcp /]# ss -anp |grep L67
[root@windows-test-dhcp /]# ss -anp |grep :67
udp                UNCONN              0                    0                                                                                      0.0.0.0%dhcp:67                  0.0.0.0:*          users:(("virt-launcher",pid=26,fd=20))   
[root@windows-test-dhcp /]# 
```

**Pull request closed without merging**